sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc

```diff
@@ -0,0 +1,364 @@
+= SYCL_INTEL_invoke_simd
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This extension introduces a mechanism to invoke explicitly vectorized functions
+written using SIMD types from a SYCL kernel.
+
+== Name Strings
+
++SYCL_INTEL_invoke_simd+
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contacts
+
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com) +
+Roland Schulz, Intel (roland 'dot' schulz 'at' intel 'dot' com) +
+Jason Sewall, Intel (jason 'dot' sewall 'at' intel 'dot' com) +
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 2 and the following extensions:
+
+- SPV_INTEL_function_pointers
+
+== Overview
+
+Standard C++ functions are treated as SPMD functions in SYCL. "Horizontal" SIMD operations (e.g.
+shuffles) can be expressed as group functions. If the user wants to tune for
+the SIMD instruction set of a particular device, then they are dependent on an
+implementation providing a sub-group extension that exposes the instructions
+that they would like to use. The user might want to express their function
+using SIMD types because they need to use intrinsics or inline ASM, have legacy
+code that uses SIMD types, or are more familiar with that programming model.
+This extension adds support for an execution model in which SIMD parallelism is
+expressed using SIMD types, with clear delineation between SPMD and SIMD code regions through an invocation interface also introduced by the extension.
+
+=== Defining a SIMD Function
+
+Uniform arguments are represented by scalars, and varying arguments by explicit
+vector types.  The explicit vector are types are based on the SIMD classes
+from [Parallelism TS2](http://wg21.link/N4808).
```

Comment:
It is not clear to me what are exact things which are imported from Parallelism TS2 and which of them are allowed to be used in SYCL SIMD functions.

I see that `std::experimental::where` usage is demonstrated in examples below, but are we going to allow the rest of functions to be also used, like `std::experimental::min`? What about math functions? Are we plan to provide functionality like `permute` or `exclusive_scan` for `sycl::ext::oneapi::simd`/`std::experimental::simd`?

Reply:
After discussing this internally we've decided that we should be aiming to implement all of the SIMD-related functions from Parallelism TS 2.

Reply:
I've tried to clarify this in 53a40f3fd670; please let me know if there are still cases that you're unsure about.

Nothing explicitly states that a function like `sycl::exclusive_scan` cannot be called in a SIMD region, but I think that's covered by the fact that there's no way to pass a group object into the SIMD region.